cmd/kubectl_token.go

```diff
@@ -381,18 +379,26 @@ func basicAuth(input *LoginInput) (managementClient.Token, error) {
 		responseType = fmt.Sprintf("%s_%s", responseType, input.clusterID)
 	}
 
-	body := fmt.Sprintf(`{"responseType":%q, "username":%q, "password":%q}`, responseType, username, password)
+	reqBody := fmt.Sprintf(`{"responseType":%q, "username":%q, "password":%q}`, responseType, username, password)
 
-	url := fmt.Sprintf("%s/v3-public/%ss/%s?action=login", input.server, input.authProvider,
+	loginURL := fmt.Sprintf("%s/v3-public/%ss/%s?action=login", input.server, input.authProvider,
 		strings.ToLower(strings.Replace(input.authProvider, "Provider", "", 1)))
 
-	response, err := request(http.MethodPost, url, bytes.NewBufferString(body))
+	req, err := http.NewRequest(http.MethodPost, loginURL, bytes.NewBufferString(reqBody))
+	if err != nil {
+		return token, fmt.Errorf("error creating request: %w", err)
+	}
+
+	resp, respBody, err := doRequest(client, req)
+	if err == nil && resp.StatusCode != http.StatusOK {
+		err = fmt.Errorf("unexpected http status code %d", resp.StatusCode)
+	}
```

Comment:
I feel having the err check before is more idiomatic, wdyt?

```go
resp, respBody, err := doRequest(client, req)
if err != nil {
	return token, nil
}
if resp.StatusCode != http.StatusOK {
	return token, fmt.Errorf("unexpected http status code %d", resp.StatusCode)
}
```

I'm also confused about the `return token, nil`. Is it ok? Should we at least log the error?

Reply:
The only idiomatic part is to explicitly check for the error if one is returned to the caller, which we do. There is no prescribed way of doing it though.
In this particular case if the request succeeds `err == nil` but the http status code is anything but the OK we want to fail and so we assign an error first L393-L395.
Then we cover both cases and check for `err != nil`.